receiver/mongodbatlasreceiver/events.go

```diff
@@ -0,0 +1,336 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbatlasreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver"
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/extension/experimental/storage"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	rcvr "go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver/internal"
+)
+
+const (
+	eventStorageKey       = "last_recorded_event"
+	defaultEventsMaxPages = 25
+	defaultEventsPageSize = 100
+	defaultPollInterval   = time.Minute
+)
+
+type eventsClient interface {
+	GetProject(ctx context.Context, groupID string) (*mongodbatlas.Project, error)
+	GetEvents(ctx context.Context, groupID string, opts *internal.GetEventsOptions) (ret []*mongodbatlas.Event, nextPage bool, err error)
+}
+
+type eventsReceiver struct {
+	client        eventsClient
+	logger        *zap.Logger
+	id            component.ID  // ID of the receiver component
+	storageID     *component.ID // ID of the storage extension component
+	storageClient storage.Client
+	cfg           *Config
+	consumer      consumer.Logs
+
+	maxPages     int
+	pageSize     int
+	pollInterval time.Duration
+	wg           *sync.WaitGroup
+	record       *eventRecord // this record is used for checkpointing last processed events
+	doneChan     chan bool
+}
+
+type eventRecord struct {
+	sync.Mutex
```

Comment:
You don't appear to be using this anywhere.

Reply:
I'm not sure you actually need the mutex. Each instance of the receiver will have its own instance of `NextStartTime`. Within each instance of the receiver, you only access `NextStartTime` synchronously within each `poll`, and each `poll` blocks the next.

Reply:
Ahh you're right I'll remove it 👍 was concerned about the goroutine nature of it but you're right it doesn't need it